### ISSUE TITLE
Add option to not use default slugify for markdown-it-anchor handling

### ISSUE THREAD

**Issue by lisastaehli** [joakin/markdown-folder-to-html#32]
Thanks for this great repo! We are currently using it to convert our markdown files (converted with https://www.npmjs.com/package/json-schema-to-markdown) of schema docs to html. We have a number of overview tables with schema properties in (sort of) camel case (fixed requirement), e.g. `GlobalID`, and underneath a list of all props with more information (e.g. type, maximum length, examples). In rendered markdown (e.g. in GitHub), when clicking on the property name in the table, it will scroll down to the more detailed description, which is super useful when having a large table of properties.

When converting to html here, `markdown-it-anchor` is used to handle these anchors which per default "slugifys" the anchor strings added to the URL (e.g. `GlobalID` --> `globalid`). Would it be possible to use an option as described in https://www.npmjs.com/package/markdown-it-anchor to overwrite the slugify default option also for markdown-folder-to-html?

```javascript
const string = require('string');
const keepStringAsIs = s => string(s);
 
const md = require('markdown-it')()
const anchor = require('markdown-it-anchor', {
    slugify: keepStringAsIs,
});
```

I am happy to look into it and make a PR if this is something that could be useful for others as well. Thanks for the consideration.

**Comment by joakin** [comment]
Hi! Thanks for opening an issue to discuss. It is cool to hear how people are using the project.

From [markdown-it-anchor/index.js#`slugify`](https://github.com/valeriangalliat/markdown-it-anchor/blob/e276fe53e259bcd2bf6045b6093f82d3cd606f8c/index.js#L1) it seems like the default slugify function doesn't do much, but one critical piece of it is `encodeURIComponent`, without it there is probably broken hashes/ids with certain text.

Before going further and overriding the default with something like the default but without the toLowerCase, would you mind posting the markdown file that is causing problems and why? I'm not sure I fully understand why the lowercase ids are a problem.

**Comment by lisastaehli** [comment]
Here's an example markdown file:

--------------------------------------

# Branches Schema

Example.

## Branches Type

`object` ([Branches](branches.md))

# Branches Properties

| Property                      | Type      | Required | Nullable       | Defined by                                                                           |
| :---------------------------- | --------- | -------- | -------------- | :----------------------------------------------------------------------------------- |
| [GlobalID](#GlobalID)         | `string`  | Optional | can be null    | [Branches](branches-properties-globalid.md "undefined#/properties/GlobalID")         |
| [CreationDate](#CreationDate) | `integer` | Optional | cannot be null | [Branches](branches-properties-creationdate.md "undefined#/properties/CreationDate") |

## GlobalID

Global ID.


`GlobalID`

-   is optional
-   Type: `string`
-   can be null
-   defined in: [Branches](branches-properties-globalid.md "undefined#/properties/GlobalID")

### GlobalID Type

`string`

## CreationDate




`CreationDate`

-   is optional
-   Type: `integer`
-   cannot be null
-   defined in: [Branches](branches-properties-creationdate.md "undefined#/properties/CreationDate")

### CreationDate Type

`integer`

-----------------------------------------

The issue is when clicking `GlobalID` or `CreationDate` in the table above, it should be scrolling down to the section GlobalID/CreationDate (imagine 20 properties in this table). With the current slugify implementation, the anchor string in the URL will be all lowercase (`#globalid`), meaning that the reference from the table (`#GlobalID`) is broken and will not scroll down to the section.

I hope this helps to understand the use case.

I agree that the `encodeURIComponent` is necessary, however the `toLowerCase` is what causes problems for us. A way forward could also be to allow passing in all the `markdown-it-anchor` config properties as a custom object, but that might be a bit more effort. Otherwise, it could just be a custom function to replace `slugify`, e.g. like this:

```javascript
const string = require('string');
const doNotUseSlugifyDefault = false; // default 
const keepStringAsIs = s => encodeURIComponent(string(s));
const defaultSlugify = encodeURIComponent(String(s).trim().toLowerCase().replace(/\s+/g, '-'));
 
const md = require('markdown-it')()
const anchor = require('markdown-it-anchor', {
    slugify: doNotUseSlugifyDefault ? keepStringAsIs : defaultSlugify,
});
```
The `doNotUseSlugifyDefault` could be passed in as an argument. Alternative would be to pass in the function itself (more flexibility) or the whole config for markdown-it-anchor (full flexibility).

Let me know what you think. If the use case is too specific, we will simply adjust the code for our purposes :). Otherwise it's working really well to create html docs.

Side note: this could of course also be fixed in the step to create the markdown files, e.g. by using slugify also create the anchor in the table to be `#globalid`.

**Comment by joakin** [comment]
Thanks for explaining more, it is very clear now.

What I'm thinking now is if it is worth deviating from upstream with a custom slugify function.

From what you comment, I don't want to add config for the reasons stated on the readme, so I consider that a non-option.

I looked at https://github.com/saibotsivad/json-schema-to-markdown/blob/master/index.js trying to see where the slugs are generated for the links, but I'm not really sure where it is happening...

Another option could be to fork json-schema-to-markdown and use the same slugify in that package for generating ids.

One final option that gets the done, although not ideal, is to use [custom HTML template](https://github.com/joakin/markdown-folder-to-html#custom-html) by copying the default template in [docs/template.html](https://github.com/joakin/markdown-folder-to-html/blob/master/docs/template.html) to your `docs` folder, and adding a small JS snippet to lowercase those table ids client side. Something like this should do it:

```js
for (let a of document.querySelectorAll('table>tbody a[href^="#"]'))
  a.href = a.href.toLowerCase()
```

This would be the quickest option and it seems fine if SEO is not important for those docs.

Let me know if one of these other options can make it work for you, and if none do, I'll make a decision about using a custom slugify function depending on the reasons.

Thanks for taking the time to chat about this! More often than not people just open PRs without discussing or asking.

**Comment by joakin** [comment]
This is probably a bit safer in case the href has case sensitive folders on it:

```js
for (let a of document.querySelectorAll('table>tbody a[href^="#"]')) {
  let url = new URL(a.href)
  a.href = url.hash.toLowerCase()
}
```

**Comment by lisastaehli** [comment]
Thanks for your detailed response and the considerations. That makes a lot of sense :) 

I just realized that I included the wrong jsonschema2md link above. We are actually using this repo here: https://github.com/adobe/jsonschema2md. I will have a look on where and how the links are parsed and generated in there. 

I will also have a look at the custom html. Thanks for the suggestion :) have a nice day!
